FSDP + DTensor state dict fix to support mix torch.Tensor and DTensor

yoyoyocmu opened the pull request:
Summary:
This is a fix to make sure that the attribute `_device_mesh` always exists before calling it.

When using together with `FSDP.state_dict_type` (i.e. during checkpointing) it is possible that the FSDP module contains both torch.tensor and DTensor.

Test Plan:
**CI**
Rely on CI test result.

**Test with WHEN model**:
Checkpoint can be saved and loaded successfully.

Differential Revision: D51419762

fegin commented:
@wz337 Do we have UTs to cover DTensor + Tensor case? I remember you encountered a similar issue before.

fegin commented:
Why device_mesh does not exist if use_dtensor is True?

yoyoyocmu commented:
It happens in the following example when we have mix torch.Tensor and DTensor in the module.
```
# FSDP authoring
module_tensor = FSDP(tensor_based_module)
module_Dtensor = FSDP(Dtensor_based_module)
roof = FSDP(module_tensor + module_Dtensor)

# For checkpointing and model publish
FSDP.set_state_dict_type(
    roof,
    state_dict_type=StateDictType.SHARDED_STATE_DICT,
)
```

yoyoyocmu committed:
FSDP + DTensor state dict fix to support mix torch.Tensor and DTensor (#113988)

Summary:

This is a fix to make sure that the attribute `_device_mesh` always exists before calling it.

When using together with `FSDP.state_dict_type` (i.e. during checkpointing) it is possible that the FSDP module contains both torch.tensor and DTensor.

Test Plan:
**CI**
Rely on CI test result.

**Test with WHEN model**:
Checkpoint can be saved and loaded successfully.

Differential Revision: D51419762

fegin commented:
This is not needed. Close the PR.
